experimental/packages/opentelemetry-sdk-metrics-base/src/export/MetricReader.ts

```diff
@@ -15,51 +15,206 @@
+ * has resolved or rejected.
+ *
+ * <p> NOTE: this operation will continue even after the timeout fires the callback.
+ *
+ * @param promise promise to use with timeout.
+ * @param timeout the timeout in milliseconds until the returned promise is rejected.
+ * @param done the callback once the promise has resolved or rejected.
+ */
+export function promiseWithTimeout<T>(promise: Promise<T>, timeout: number, done: (result: ReaderResult<T>) => void): void {
+  // keep handle so that we can clear it later.
+  let timeoutHandle: ReturnType<typeof setTimeout>;
+
+  // Set up a promise to handle the timeout.
+  const timeoutPromise = new Promise<ReaderResult<T>>(function timeoutFunction(resolve) {
+    timeoutHandle = setTimeout(
+      function timeoutHandler() {
+        resolve({
+          code: ReaderResultCode.TIMED_OUT,
+          error: new Error('Operation timed out.')
+        })
+      },
+      timeout
+    );
+  });
+
+  // Wrap to promise to get a result code with the result if it does not reject.
+  const resultCodePromise = promise.then(result => {
+    return { code: ReaderResultCode.SUCCESS, returnValue: result }
+  })
+
+  Promise.race([resultCodePromise, timeoutPromise]).then(result => {
+      // Clear timeout on success and return result.
+      clearTimeout(timeoutHandle);
+      if (done) {
+        done(result);
+      }
+    },
+    reason => {
+      // Clear timeout on rejection and return failure.
+      clearTimeout(timeoutHandle);
+      if (done) {
+        done({
+          code: ReaderResultCode.FAILED,
+          error: reason
+        });
+      }
+    });
+}
 
 // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader
 
+/**
+ * A registered reader of metrics that, when linked to a {@link MetricProducer}, offers global
+ * control over metrics.
+ */
 export abstract class MetricReader {
+  // Tracks the shutdown state.
   private _shutdown = false;
+  // MetricProducer used by this instance.
   private _metricProducer?: MetricProducer;
 
-  constructor(private _exporter: MetricExporter) {}
+  constructor(private readonly _preferredAggregationTemporality = AggregationTemporality.CUMULATIVE) {
+  }
 
+  /**
+   * Set the {@link MetricProducer} used by this instance.
+   *
+   * @param metricProducer
+   */
   setMetricProducer(metricProducer: MetricProducer) {
     this._metricProducer = metricProducer;
+    this.onInitialized();
   }
 
+  /**
+   * Get the {@link AggregationTemporality} preferred by this {@link MetricReader}
+   */
   getPreferredAggregationTemporality(): AggregationTemporality {
-    return this._exporter.getPreferredAggregationTemporality();
+    return this._preferredAggregationTemporality;
   }
 
-  async collect(): Promise<void> {
+  /**
+   * Handle once the SDK has initialized this {@link MetricReader}
+   *
+   * @protected
+   */
+  protected abstract onInitialized(): void;
```

Comment:
Can we provide a default implementation for non-necessarily overridden methods? Like, say in that way we can reduce the verbosity in the following tests.